lib/semantics/resolve-labels.cc

```diff
@@ -0,0 +1,917 @@
+            currentScope_, currentPosition_, labeledStmtClassificationSet})};
+    if (!pair.second) {
+      errorHandler_.Say(currentPosition_,
+          parser::MessageFormattedText{
+              "label '%" PRIu64 "' is not distinct"_err_en_US, label});
+    }
+  }
+
+  void addLabelReferenceFromDoStmt(parser::Label label) {
+    CheckLabelInRange(label);
+    programUnits_.back().doStmtSources.emplace_back(
+        label, currentScope_, currentPosition_);
+  }
+
+  void addLabelReferenceFromFormatStmt(parser::Label label) {
+    CheckLabelInRange(label);
+    programUnits_.back().formatStmtSources.emplace_back(
+        label, currentScope_, currentPosition_);
+  }
+
+  void addLabelReference(parser::Label label) {
+    CheckLabelInRange(label);
+    programUnits_.back().otherStmtSources.emplace_back(
+        label, currentScope_, currentPosition_);
+  }
+
+  void addLabelReference(const std::list<parser::Label> &labels) {
+    for (const parser::Label &label : labels) {
+      addLabelReference(label);
+    }
+  }
+
+  std::vector<UnitAnalysis> programUnits_;
+  parser::Messages errorHandler_;
+  parser::CharBlock currentPosition_{nullptr};
+  ProxyForScope currentScope_{0};
+  std::vector<std::string> constructNames_;
+};
+
+bool InInclusiveScope(const std::vector<ProxyForScope> &scopes,
+    ProxyForScope tail, ProxyForScope head) {
+  for (; tail != head; tail = scopes[tail]) {
+    if (!HasScope(tail)) {
+      return false;
+    }
+  }
+  return true;
+}
+
+ParseTreeAnalyzer LabelAnalysis(const parser::Program &program) {
+  ParseTreeAnalyzer analysis;
+  Walk(program, analysis);
+  return analysis;
+}
+
+bool InBody(const parser::CharBlock &position,
+    const std::pair<parser::CharBlock, parser::CharBlock> &pair) {
+  if (position.begin() >= pair.first.begin()) {
+    if (position.begin() < pair.second.end()) {
+      return true;
+    }
+  }
+  return false;
+}
+
+LabeledStatementInfoTuplePOD GetLabel(
+    const TargetStmtMap &labels, const parser::Label &label) {
+  const auto iter{labels.find(label)};
+  if (iter == labels.cend()) {
+    return {0u, nullptr, LabeledStmtClassificationSet{}};
+  } else {
+    return iter->second;
+  }
+}
+
+// 11.1.7.3
+void CheckBranchesIntoDoBody(const SourceStmtList &branches,
+    const TargetStmtMap &labels, const std::vector<ProxyForScope> &scopes,
+    const IndexList &loopBodies, parser::Messages &errorHandler) {
+  for (const auto branch : branches) {
+    const auto &label{branch.parserLabel};
+    auto branchTarget{GetLabel(labels, label)};
+    if (HasScope(branchTarget.proxyForScope)) {
+      const auto &fromPosition{branch.parserCharBlock};
+      const auto &toPosition{branchTarget.parserCharBlock};
+      for (const auto body : loopBodies) {
+        if (!InBody(fromPosition, body) && InBody(toPosition, body)) {
+          errorHandler.Say(fromPosition,
+              parser::MessageFormattedText{
+                  "branch into '%s' from another scope"_en_US,
```

Comment:
`scope` might be better if replaced with `outside the body of the loop` in this message.